### PR TITLE
Latency reporting for LV2

### PR DESCRIPTION
**Pull request by jpcima**
#469
This adds latency reporting for LV2.
A control port will be added for this special purpose. It does not show on the host UI.
The port is created only if there exists audio output ports.

Observation: under the Ardour host, the experiment consists of 2 identical audio tracks, except one equipped with a Dplug LV2 plugin which delays by `N`.
- the normal track and delayed track are playing in sync
- when the `N` is varied dynamically by parameter, Ardour adjusts dynamically and the playback gets back on its feet after a moment

**Note** given that latency reporting is dynamic, behaviour may differ from other clients.
It's also the case that `latencySamples` is invoked from the `run` thread.

Attached is the source of the plugin for test. Not included in commit because it's kinda useless outside this.
[latency-introducer.tar.gz](https://github.com/AuburnSounds/Dplug/files/5724172/latency-introducer.tar.gz)

Some things to note at LV2 level:
- the simplifies a bit the port numbering gymnastics. This is welcome in case there are optionals which introduce gaps, such as the latency port.
- the LV2 generator has some problems which needed solving; for instance, the outputs were labelled from 0, but inputs were labelled according to their starting port index.
- other problem, in the exporter, a terminating `.` character would be introduced in some conditions for intermediate ports, but it cannot be because there is always a next port (events port is last, and always existing)

### CODE REVIEW

**Comment by p0nce** [comment]
Will review tomorrow!

**Comment by p0nce** [comment]
**tl;dr will be merged when your reintroduce former lv2:symbol from pre-patch**


Look good except the showstopper:

> the LV2 generator has some problems which needed solving; for instance, the outputs were labelled from 0, but inputs were labelled according to their starting port index.

Changing this is actually a problem ! 

image

Changing the `lv2:symbol` breaks user sessions. Quoting http://lv2plug.in/ns/lv2core
Dplug tries to never, ever break user sessions unless choosen by plugin developer.

We have too much LV2 plugins in the wild.

> A port has two identifiers: a (numeric) index, and a (textual) symbol. The index can be used as an identifier at run-time, but persistent references to ports (for example in presets or save files) MUST use the symbol. Only the symbol is guaranteed to refer to the same port on all plugins with a given URI, that is the index for a port may differ between plugin binaries.

Something that break user sessions must introduce an Option https://github.com/AuburnSounds/Dplug/wiki/Options
In this case I don't think it's worth the hassle.
So to be merged you would have to reintroduce the ugly gymnatics that was there before...

(On a completely unrelated note: `latencySamples` must purely depends from sample rate and cannot depend from a Parameter. Explanations here => https://github.com/AuburnSounds/Dplug/issues/442 of course this doesn't affect mergeability)

**Comment by jpcima** [comment]
Good remark yes, changing the symbol is going to affect the compatibility.
But there's a bigger dilemma here actually.

Originally, the input port symbol is offset by the parameter count.
The implication is, if preservation of compatibility is desired, one can never add any parameters!
The present state of lv2 will have a problem to stand in the long term.

>(On a completely unrelated note: latencySamples must purely depends from sample rate and cannot depend from a Parameter. Explanations here => #442 of course this doesn't affect mergeability)

In this case it should compute the value just once on plugin activation.

**Comment by p0nce** [comment]
Regarding adding parameters: I though this was allowed by formats like VST2 / VST3 but at times I've broken user sessions doing this and got some really heart-breaking emails.

So I've updated https://github.com/AuburnSounds/Dplug/wiki/Dplug-Save-Compatibility to say:
> You may try to add a Parameter at the end of the list of parameters constructed in buildParams(). It is currently unknown if this keeps compatibility in all DAW. Avoid if you can.

The reality is that we have no evidence that we can add parameters in any other formats, even though it was previously assumed one could by adding at the end of the list.

It will be good if we allow to add Parameters in LV2 of course! But this still must be under an Option since existing plugins take precedence (ideally in another issue, they are unrelated)

**Comment by jpcima** [comment]
This last commit reintroduces the backward compatibility.
Not checked yet, I'll run a diff of all plugins to check all went fine.

I'll do fixed latency next a give this a test run in Ardour DAW.

**Comment by p0nce** [comment]
Diff between manifest is 100% OK here!

After more reviewing: minor point that you might or might not fix:
- [ ] The value `_client.latencySamples(_sampleRate);` can be cached into a field and set in `instantiate`
```
    if (_latencyOutput)
            *_latencyOutput = _client.latencySamples(_sampleRate); 
```

- [ ] Make the latency port always there (like the "events" input), which simplifies the generation of the manifest`with the regards to the line ` if(output == legalIO.numOutputChannels - 1)` it took me a while to see why it was in the loop.

But if you prefer I can merge as is. (you have earned the bounty either way)

**Comment by jpcima** [comment]
>  The value _client.latencySamples(_sampleRate); can be cached into a field and set in instantiate

that yes, I did in 38ab9b3 and just checked in Ardour, it worked fine

> Make the latency port always there

I was told that in LV2 it's preferable to limit the number of control ports, for performance. (by R Gareus from Ardour)
A reason for that is that the DAW has to poll each one to scan for value changes.
It's true though, 1 port does rarely make a difference, 100s of them can.
LV2 has been working to change their model towards get/set messages like VST uses. (the patch messages)

**Comment by p0nce** [comment]
OK right.
```
<rgareus> p0nce: It makes sense to report it for every plugin that has a an input (audio or MIDI) and an output (audio or MIDI).
```
Dplug doesn't have MIDI output yet so it's OK that way!
